### PR TITLE
Describe record cursor type mapping in dev guide

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

Describe how to map types when implementing a ``RecordCursor``, like how to encode strings into slices or pack timestamps. Add some basic examples of creating blocks.

> Is this change a fix, improvement, new feature, refactoring, or other?
improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
docs

> How would you describe this change to a non-technical end user or system administrator?
n/a

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by colebow — COMMENTED** [review]
This overall looks good to me, though I definitely don't know enough about the subject matter to review for correctness / accuracy.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 1]
```diff
@@ -38,20 +38,114 @@ connector does), you may need to get creative about how you map your
...
+pushing down predicates, limit or top-N, and informs the engine
+that this is not necessary.
+
+For simple connectors, where performance is not critical,
+the record set provider can return an instance of ``InMemoryRecordSet``.
+It can be built using lists of values for every row, which can be
+simpler than implementing a ``RecordCursor``.
+
+A ``RecordCursor`` implementation needs to keep track of the current record,
+and know how to return values for columns by a numerical position, in
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks, I broke them down.

**Comment by MiguelWeezardo** [comment]
Looks good

File: docs/src/main/sphinx/develop/connectors.rst [hunk 2]
```diff
@@ -82,30 +82,122 @@ a :ref:`connector-page-sink-provider`.
...
+A ``RecordCursor`` implementation needs to keep track of the current record.
+It return values for columns by a numerical position, in the data type matching
+the column definition in the table. When the engine is done reading the current
+record it will call ``advanceNextPosition`` on the cursor.
+
+Type mapping
+^^^^^^^^^^^^
+
+Both the ``InMemoryRecordSet`` and the ``RecordCursor`` use the following
+Java types:
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:126]: The Java type needs to match the carrier type of the corresponding SQL type being returned. 

The list below is incomplete and slightly misleading. It would be better to describe the SQL types, the carrier types and then explain how they relate to the methods in RecordCursor.

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks! I changed this into a table, listing all built-in SQL types and their carrier types. PTAL. For `HyperLogLog` and others I left it as `Object`, do you know what are their carrier types or where to find them?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 3]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+Type mapping
+^^^^^^^^^^^^
+
+The built-in SQL data types use different Java types as carrier types.
+
+================================== =============
+SQL type                           Java type
+================================== =============
+``BOOLEAN``                        ``boolean``
+``TINYINT``                        ``long``
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:131]: When using `InMemoryRecordSet` I could use a `byte` for this. Does `RecordCursor` have to return its value in `getLong` or can it return a byte in `getObject`?

**Comment by nineinchnick** [comment]
@martint PTAL

File: docs/src/main/sphinx/develop/connectors.rst [hunk 4]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+--------------------------
 
 Given a split and a list of columns, the record set provider is
 responsible for delivering data to the Trino execution engine.
+
 It creates a ``RecordSet``, which in turn creates a ``RecordCursor``
 that is used by Trino to read the column values for each row.
 
+The provided record set must only include requested columns
+in the correct order. If it returns more rows than necessary,
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:107]: "the correct order" is too vague. Make it more explicit by explaining how it relates to the order of the columns in the `getRecordSet()` call.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 5]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+the engine will filter them out unless the connector implements
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:108]: This is misleading. The engine *may* filter, if the query so requires it. But a `SELECT * FROM t` will not filter anything.

The connector needs to return all the rows contained in the "virtual table" represented by the TableHandle associated with the TableScan operation. It shouldn't return more or fewer rows.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 3 above)
Comment (martint) [reply docs/src/main/sphinx/develop/connectors.rst:131]: What do you mean by "I could use a `byte`"? For types that expect a `long` carrier, the RecordCursor should implement `getLong`. `getObject` will not get called in that case. See `RecordPageSource.getNextPage()`

File: docs/src/main/sphinx/develop/connectors.rst [hunk 6]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+                                   than 6
+``TIMESTAMP(P) WITH TIME ZONE``    ``long`` for precision up to 3;
+                                   ``LongTimestampWithTimeZone`` for precision
+                                   greater than 3
+``INTERVAL YEAR TO MONTH``         ``long``
+``INTERVAL DAY TO SECOND``         ``long``
+``ARRAY``                          ``Block``
+``MAP``                            ``Block``
+``ROW``                            ``Block``
+``IPADDRESS``                      ``[]byte``
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:159]: byte[]

However, the carrier type for IPADDRESS and UUID is Slice

File: docs/src/main/sphinx/develop/connectors.rst [hunk 7]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+the carrier type:
+
+* ``getBoolean(int field)``
+* ``getLong(int field)``
+* ``getDouble(int field)``
+* ``getSlice(int field)``
+* ``getObject(int field)``
+
+
+Date and time values of regular precision can be converted into ``long`` using
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:179]: This is only for `timestamp(p) with time zone` and `time(p) with time zone` types.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 8]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+the ``Slices.utf8Slice()`` static method.
+
+.. note::
+
+  ``Slice`` class is provided by the ``io.airlift:slice`` package.
+
+``Int128`` objects for decimal columns can be created using
+the ``Int128.valueOf()`` method,
+
+Example of creating a block for an ``ARRAY(VARCHAR)``  column:
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:193]: Lower case for type names here and everywhere else: `array(varchar)`

File: docs/src/main/sphinx/develop/connectors.rst [hunk 9]
```diff
@@ -82,30 +82,158 @@ a :ref:`connector-page-sink-provider`.
...
+like ``pack()`` or ``packDateTimeWithZone()``.
+
+UTF-8 encoded strings can be converted to Slices using
+the ``Slices.utf8Slice()`` static method.
+
+.. note::
+
+  ``Slice`` class is provided by the ``io.airlift:slice`` package.
+
+``Int128`` objects for decimal columns can be created using
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:190]: Int128 is not just for decimal columns (even though it's currently only used by that type). It's for anything that needs a 128-bit integer number.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 3 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:131]: That answers my question, thanks!

File: docs/src/main/sphinx/develop/connectors.rst (hunk 4 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks! I added a sentence about the list of column handles passed to `getRecordSet()`.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 5 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: I used your second paragraph almost verbatim. I wonder if we can make it more explicit to readers that they have complete control over what's included in the `TableHandle`, but we can iterate on this.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 6 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks! I used `InMemoryRecord` set too much :-)

File: docs/src/main/sphinx/develop/connectors.rst (hunk 8 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Is it ok to keep them uppercase in the table? I based it on the [data types](https://trino.io/docs/current/language/types.html) section.

**Comment by nineinchnick** [comment]
@martint AC, PTAL, and thanks so much!

**Comment by nineinchnick** [comment]
@martint PTAL, there are two comments I think I addressed but need your confirmation
@mosabua PTAL generally (or maybe you know others who could review the wording)

File: docs/src/main/sphinx/develop/connectors.rst [hunk 10]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+record it will call ``advanceNextPosition`` on the cursor.
+
+Type mapping
+^^^^^^^^^^^^
+
+The built-in SQL data types use different Java types as carrier types.
+
+================================== =============
+SQL type                           Java type
+================================== =============
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:275]: Please replace this table format with list-table. Here's an example: https://github.com/trinodb/trino/pull/13118/files

File: docs/src/main/sphinx/develop/connectors.rst [hunk 11]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
 .. _connector-record-set-provider:
 
 ConnectorRecordSetProvider
-^^^^^^^^^^^^^^^^^^^^^^^^^^
+--------------------------
+
+Given a split and a list of columns, the record set provider is responsible for
+delivering data to the Trino execution engine. It creates a ``RecordSet``, which
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:244]: In general, avoid using "it" to remove ambiguity about the subject. In this case, I'd suggest "The provider creates..."

File: docs/src/main/sphinx/develop/connectors.rst [hunk 12]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+in turn creates a ``RecordCursor`` that's used by Trino to read the column
+values for each row.
+
+The provided record set must only include requested columns in the order
+matching the list of column handles passed to the
+``ConnectorRecordSetProvider.getRecordSet()`` method. It must return all the
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:250]: Replace "it" with "The record set" or similar

File: docs/src/main/sphinx/develop/connectors.rst [hunk 13]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+rows contained in the "virtual table" represented by the TableHandle associated
+with the TableScan operation. It shouldn't return more or fewer rows.
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:252]: Replace "It" with "This set" or similar

Comment (jhlodin) [reply docs/src/main/sphinx/develop/connectors.rst:252]: "should return the exact same number of rows" is clearer to me than "shouldn't return more or fewer"

File: docs/src/main/sphinx/develop/connectors.rst [hunk 14]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+
+The engine may filter returned rows. The connector can implement pushing down
+predicates, limit, or top-N, to informs the engine what filtering is not
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:255]: to inform

Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:255]: Suggest rewording to "can implement predicate, limit, or top-N pushdown to inform"

File: docs/src/main/sphinx/develop/connectors.rst [hunk 15]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+necessary.
+
+For simple connectors, where performance is not critical, the record set
+provider can return an instance of ``InMemoryRecordSet``. It can be built using
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:259]: Replace "It" with "The record set" or similar

File: docs/src/main/sphinx/develop/connectors.rst [hunk 16]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+lists of values for every row, which can be simpler than implementing a
+``RecordCursor``.
+
+A ``RecordCursor`` implementation needs to keep track of the current record.
+It return values for columns by a numerical position, in the data type matching
+the column definition in the table. When the engine is done reading the current
+record it will call ``advanceNextPosition`` on the cursor.
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:266]: Our written style is to avoid future tense. Suggest using "it calls" instead

File: docs/src/main/sphinx/develop/connectors.rst [hunk 17]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+The ``RecordCursor.getType(int field)`` method returns the SQL type for a field
+and the value for it is returned by one of the following methods, matching
+the carrier type:
+
+* ``getBoolean(int field)``
+* ``getLong(int field)``
+* ``getDouble(int field)``
+* ``getSlice(int field)``
+* ``getObject(int field)``
+
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:323]: Remove extra line

File: docs/src/main/sphinx/develop/connectors.rst [hunk 18]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+UTF-8 encoded strings can be converted to Slices using
+the ``Slices.utf8Slice()`` static method.
 
-Given a split and a list of columns, the record set provider is
-responsible for delivering data to the Trino execution engine.
-It creates a ``RecordSet``, which in turn creates a ``RecordCursor``
-that's used by Trino to read the column values for each row.
+.. note::
+
+  ``Slice`` class is provided by the ``io.airlift:slice`` package.
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:335]: The ``Slice`` class

File: docs/src/main/sphinx/develop/connectors.rst [hunk 19]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+
+``Int128`` objects can be created using the ``Int128.valueOf()`` method.
+
+Example of creating a block for an ``array(varchar)``  column:
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:339]: "The following example creates a block for..."

File: docs/src/main/sphinx/develop/connectors.rst [hunk 20]
```diff
@@ -238,17 +238,148 @@ the strategy employed by the Example HTTP connector.
...
+                builder.appendNull();
+            }
+            else {
+                VARCHAR.writeString(builder, name);
+            }
+        }
+        return builder.build();
+    }
+
+Example of creating a block for a ``map(varchar, varchar)`` column:
```
Comment (jhlodin) [inline docs/src/main/sphinx/develop/connectors.rst:357]: "The following example creates a block for..."

**Review by jhlodin — COMMENTED** [review]
Some suggested style and formatting edits

File: docs/src/main/sphinx/develop/connectors.rst (hunk 10 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Done!

File: docs/src/main/sphinx/develop/connectors.rst (hunk 13 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: I dropped the last sentence as in this form it would just duplicate the previous one.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 16 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Ah, this is something that the Vale linter caught! I just started using it.

Comment (mosabua) [reply docs/src/main/sphinx/develop/connectors.rst]: We are still requesting a demo of that from you some time soon btw @nineinchnick

**Comment by nineinchnick** [comment]
@martint PTAL

File: docs/src/main/sphinx/develop/connectors.rst [hunk 21]
```diff
@@ -238,17 +238,174 @@ the strategy employed by the Example HTTP connector.
...
+        return builder.build();
+    }
+
+The following example creates a block for a ``map(varchar, varchar)`` column:
+
+.. code-block:: java
+
+    private Block encodeMap(Map<String, ?> map)
+    {
+        MapType mapType = new MapType(VARCHAR, VARCHAR, new TypeOperators());
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:389]: The type should be created via `TypeManager.getParameterizedType()`

File: docs/src/main/sphinx/develop/connectors.rst [hunk 22]
```diff
@@ -238,17 +238,174 @@ the strategy employed by the Example HTTP connector.
...
+        BlockBuilder values = mapType.createBlockBuilder(null, map != null ? map.size() : 0);
+        if (map == null) {
+            values.appendNull();
+            return values.build().getObject(0, Block.class);
+        }
+        BlockBuilder builder = values.beginBlockEntry();
+        for (Map.Entry<String, ?> entry : map.entrySet()) {
+            VARCHAR.writeString(builder, entry.getKey());
+            Object value = entry.getValue();
+            VARCHAR.writeString(builder, value != null ? value.toString() : "");
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:399]: This is not equivalent. It's converting null values to empty strings. It should use `builder.appendNull()` instead.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 21 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Done!

File: docs/src/main/sphinx/develop/connectors.rst (hunk 22 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks for cathing this, done!

**Comment by nineinchnick** [comment]
@martint there are still two unresolved comments where I asked for clarification, PTAL.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 2 above)
Comment (martint) [reply docs/src/main/sphinx/develop/connectors.rst]: You need to look at the output of `Type.getJavaType()`

File: docs/src/main/sphinx/develop/connectors.rst (hunk 8 above)
Comment (martint) [reply docs/src/main/sphinx/develop/connectors.rst]: Yes, that's ok.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 2 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks, updated!

**Comment by nineinchnick** [comment]
@martint all done!
